Red Border for Required Plugins

cyberwani opened the pull request:
added left-border in plugin list for required plugins.

cyberwani committed:
Red Border for Required Plugins

added left-border in plugin list for required plugins.

GaryJones commented:
Thanks @cyberwani.

Can you please show a screenshot with your change?
Have you checked what happens with the border when the screen is narrow and lines wrap around?
Is the red the same hex code of the attention grabbing red in WP core? (For which there's a Trac ticket open by @afercia) to consolidate these down).

You'll need to ensure you meet code standards - see https://travis-ci.org/TGMPA/TGM-Plugin-Activation/jobs/159883749 for issues with your PR.

cyberwani commented:
Here are some screenshots:
![screenshot-320px](https://cloud.githubusercontent.com/assets/1647283/18575969/8fa0602a-7bfa-11e6-90f2-695c91a58b9b.png)
![screenshot-480px](https://cloud.githubusercontent.com/assets/1647283/18575970/8fa227de-7bfa-11e6-9172-532e7d978a9b.png)
![screenshot-768px](https://cloud.githubusercontent.com/assets/1647283/18575968/8f9fc80e-7bfa-11e6-8d60-aa2b77772d70.png)
![screenshot-1440px-cropped](https://cloud.githubusercontent.com/assets/1647283/18575971/8fa55058-7bfa-11e6-9165-1497e9dd189a.png)

GaryJones commented:
Thanks for providing those. Visually, I'm OK with that. @jrfnl ?

@cyberwani The code will still need fixing for code standards.

jrfnl commented:
Visually I like the change, nice one!

Code-wise, quite apart from the code style issues, I'd like to properly review it before merging and it may be a little while before I have the time.

cyberwani commented:
@GaryJones @jrfnl I am not familiar with Travis. So, can you both suggest me the further steps?

jrfnl commented:
Travis is an automated code check which runs on the code and causing one of the red crosses below.
If you click on "Details", you can see which checks are run and which is failing.
In this case it is failing as the code does not comply with the [WordPress coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
For details of what you need to fix, click on the failing build to see the feedback given there: 
https://travis-ci.org/TGMPA/TGM-Plugin-Activation/jobs/159883749#L202

GaryJones commented:
For reference, the following code standards (formatting of the code) need to be addressed:

```
----------------------------------------------------------------------
FOUND 6 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 2099 | ERROR | [ ] Missing function doc comment
      |       |     (Squiz.Commenting.FunctionComment.Missing)
 2754 | ERROR | [x] Concat operator must be surrounded by a single
      |       |     space
      |       |     (Squiz.Strings.ConcatenationSpacing.PaddingFound)
 2754 | ERROR | [ ] Expected next thing to be an escaping function (see
      |       |     Codex for 'Data Validation'), not 'strtolower'
      |       |     (WordPress.XSS.EscapeOutput.OutputNotEscaped)
 2754 | ERROR | [x] Expected 1 spaces after opening bracket; 0 found
      |       |     (PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket)
 2754 | ERROR | [x] Expected 1 spaces before closing bracket; 0 found
      |       |     (PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket)
 2754 | ERROR | [x] Concat operator must be surrounded by a single
      |       |     space
      |       |     (Squiz.Strings.ConcatenationSpacing.PaddingFound)
----------------------------------------------------------------------
```

So, for the indicated lines:
1. Add a full DocBlock in the correct format for the new function you added.
2. Add a space either side of the `.`.
3. Add `esc_attr()` call around the `strtolower()` call.
4, 5 and 6. Add extra spaces as indicated.

cyberwani committed:
Code Formatting Issues Fixed

Code formatting issue mentioned in https://travis-ci.org/TGMPA/TGM-Plugin-Activation/jobs/159883749#L202

cyberwani committed:
Code Formatting Issues Fixed

https://travis-ci.org/TGMPA/TGM-Plugin-Activation/jobs/161903675

GaryJones commented on class-tgm-plugin-activation.php:
```diff
@@ -2096,6 +2096,13 @@ public static function get_instance() {
 		}
 
+		/**
+		 * Adds css to admin head
```
"CSS" in capitals please.
Add a full stop at the end of the description.

cyberwani replied:
Above request fixed.

GaryJones commented on class-tgm-plugin-activation.php:
```diff
@@ -2096,6 +2096,13 @@ public static function get_instance() {
+		 *
+		 * @since 2.6.2
+		 *
+		 * @return none
```
Leave this `@return` line (and the empty line above it) out please.

GaryJones commented on class-tgm-plugin-activation.php:
```diff
@@ -2093,6 +2095,19 @@ public static function get_instance() {
+		 *
+		 * @since 2.6.2
+		 */
+		function tgm_css() {
```
I'd like this to be renamed to `admin_css` (and the hook reference), just to be a little more specific.

cyberwani replied:
Changed name ad described above.

GaryJones commented on class-tgm-plugin-activation.php:
```diff
@@ -451,6 +451,8 @@ public function init() {
 				add_action( 'switch_theme', array( $this, 'force_deactivation' ) );
 			}
+
+			add_action( 'admin_head',  array( $this, 'tgm_css' ), 21 );
```
Where has this magic `21` come from? Maybe a comment is needed to explain why that priority was chosen.

cyberwani replied:
It was set to 21 as because if someone wants to add other CSS formatting with low or higher priority. But, I removed the priority.

GaryJones commented on class-tgm-plugin-activation.php:
```diff
@@ -2093,6 +2095,19 @@ public static function get_instance() {
+		 */
+		function tgm_css() {
+			echo '<style>
+			#tgmpa-plugins #the-list .required > th {
```
Do we really need two IDs in the selector? At most, one should be needed, but ideally, none if classes are available.

cyberwani replied:
Removed double ID, now just using #tgmpa-plugins.

GaryJones commented on class-tgm-plugin-activation.php:
```diff
@@ -2741,7 +2756,9 @@ protected function get_row_actions( $item ) {
 		 */
 		public function single_row( $item ) {
-			parent::single_row( $item );
+			echo '<tr class="' . esc_attr( strtolower( $item['type'] ) ) . '">';
```
This `required` class is too generic - it may well clash with other plugins or maybe even core at some point in the future. Let's go for something like:

```
echo '<tr class="' . esc_attr( 'tgmpa-type-' . strtolower( $item['type'] ) ) . '">';
```

...and updating the CSS selector to match.
(With a prefixed class name, that may even allow the specificity of the selector to be dropped right down.)

cyberwani replied:
Yes, it may be possible, added provided prefix.

cyberwani committed:
Requested changes fixed.

jrfnl commented on class-tgm-plugin-activation.php:
```diff
@@ -2093,6 +2095,19 @@ public static function get_instance() {
+		 *
+		 * @since 2.6.2
+		 */
+		function admin_css() {
```
May I suggest moving this method up a little ? To just above the `get_instance()` method which should be the last in the class ?

cyberwani replied:
Yes, it's moved now.

jrfnl commented on class-tgm-plugin-activation.php:
```diff
@@ -2741,7 +2756,9 @@ protected function get_row_actions( $item ) {
 		public function single_row( $item ) {
-			parent::single_row( $item );
+			echo '<tr class="' . esc_attr( 'tgmpa-type-' . strtolower( $item['type'] ) ) . '">';
+			parent::single_row_columns( $item );
```
Please change this to `$this->single_row_columns( $item )` to allow for TGMPA potentially adding that method to the child class.

cyberwani replied:
Changed.

jrfnl commented on class-tgm-plugin-activation.php:
```diff
@@ -2093,6 +2095,19 @@ public static function get_instance() {
+		 * @since 2.6.2
+		 */
+		function admin_css() {
+			echo '<style>
```
I'd suggest starting with a new line, so this code will present aligned in the HTML,

``` php
    echo '
    <style>
```

And maybe indent the css nicely as well ?

cyberwani replied:
Changed as above.

jrfnl commented on class-tgm-plugin-activation.php:
```diff
@@ -451,6 +451,8 @@ public function init() {
 				add_action( 'switch_theme', array( $this, 'force_deactivation' ) );
 			}
+
+			add_action( 'admin_head',  array( $this, 'admin_css' ) );
```
Please add a comment above this line to indicate why this hook is there - something along the lines of `// Add CSS for the TGMPA admin page.`

cyberwani replied:
Added.

jrfnl commented on class-tgm-plugin-activation.php:
```diff
@@ -2093,6 +2095,19 @@ public static function get_instance() {
+		 *
+		 * @since 2.6.2
+		 */
+		function admin_css() {
```
The CSS should only be added when the TGMPA page is being loaded, not for any other admin pages.

``` php
            if ( ! $this->is_tgmpa_page() ) {
                return;
            }
```

cyberwani replied:
Added.

cyberwani committed:
Requested changes fixed.
